### ISSUE TITLE
Map CoreMedia locales to supported locales

### ISSUE THREAD

**Issue by mtommila** [CoreMedia/coremedia-globallink-connect-integration#31]
The site structure in CoreMedia might be slightly different from the supported locales on gcc side.

For example, we have a two-level site structure where we have a "global master" site, the World English site. From that site we derive all country-specific English language sites such as the Germany English site (with locale en-DE). From the Germany English site we then derive the Germany German site (with locale de-DE).

Similarly e.g. for Austria we have an Austria English site (en-AT) and from that derived the Austria German site (de-AT).

For the German language, we only want to do the translation once, to de-DE. We don't separately want to translate de-AT, we want de-AT to be just a copy of de-DE.

Additionally we have in our system some other websites, that just have the locale NO_COUNTRY German, i.e. locale "de". This we also want to translate with the locale de-DE on gcc side.

Thus what we need is some way to map the source and target locales, before sending content to translation. For example if we translate the Austria German site, we want the source locale to be en-GB (instead of en-AT, which does not really even exist as a language). Similarly we want the target locale to be de-DE (instead of de-AT, which we don't even have configured on gcc side).

It should be rather straightforward to do this mapping in the SendToGlobalLinkAction class. It could be configurable with a boolean property if this mapping should be done or not. The supported locales can be easily retrieved from the gcc API with getConnectorsConfig().getSupportedLocales(). The best matching supported locale could be found by primarily matching the language and secondarily the country (if there are several matches for the language, or if there are no matches for the language like what could be the case in mapping no-NO to nb-NO).

**Comment by aprantl** [comment]
Thanks for the suggestion @mtommila. Afaik changing the locales of the source and submitted content could lead to unexpected results during XLIFF import. The import and export are also customizable so it could change the behavior of existing clients. 

I am not sure though and as such I will discuss and investigate with @mmichaelis. Our test resources are limited and I am not sure yet how to properly test this path.

**Comment by mtommila** [comment]
Hello,

I was thinking about the XLIFF import part but the "invalid locale" XLIFF import error is simply ignored so it should not be a problem. In fact I have tested this many times on my machine and it works fine. Please see the the implementation in Pull Request #32.

In my implementation this feature is by default disabled so it doesn't do anything unless you enable it. Existing clients should not be affected.

It could be tested e.g. with the following setup:
Site locales:
* master: en-US
* derived: de-AT, no-NO, zh-CN, zh-HK

Supported locales on gcc side:
* source: en-GB
* target: de-DE, nb-NO, zh-CN, zh-HK

Best regards,
Mikko

**Comment by mtommila** [comment]
And if you still conclude that this is unacceptable, what I would only need at the very minimum is that the method `SendToGlobalLinkAction.submitSubmission()` is made protected so I could override it in a subclass it and add the functionality that I need.

Best regards,
Mikko

**Comment by mmichaelis** [comment]
Hi @mtommila,

thanks for suggesting. This may require some more research, though. We have an internal issue (CMS-18577 for reference) which seems to be at least related to your request. It's entitled: _Quirk XLIFF-Export vs. -Import and Locales: Cannot fully adapt locale-representation_.

If I got you right, the XLIFF will contain "invalid" locales from the CMS perspective. While this may be ignored on import (I read our issue differently, but that's to be tested), links in CoreMedia Richtext will be broken after trying such import (again, derived from our issue description, may need to be re-investigated).

The feature you describe reads similar to an early requirement we got on certification phase, which was to provide locale mapping within CoreMedia CMS. After consulting TDC back in 2019 it turned out, that the mapping of locales should be completely maintained in PD, and it would not be up to the connected CMS to do the mapping.

In other words: We have to investigate here again and get in touch with TDC, if requirements changed meanwhile.

**Comment by mtommila** [comment]
This sounds interesting, I will ask Transperfect about mapping the locales on their side. If that works, that would work for us just as well. Strangely, this never came up in the meetings we had with them.

**Comment by mtommila** [comment]
I asked Transperfect and indeed the locale mapping can be done on Project Director side. So it's not necessary to do that on CoreMedia side.

Having the `SendToGlobalLinkAction.submitSubmission()` method protected might still be useful in the future for unknown future needs.
